### PR TITLE
Revit_Toolkit: Query the top/bottom centreline of an IFramingElement

### LINKED ISSUE #894: Revit_Toolkit: Query the top/bottom centreline of an IFramingElement based on Revit parameter
#### Description:

Whether the location line on an IFramingElement in Revit defines the centre top constraint of the element, or the centre centre or the centre bottom is controlled by the "z Justification" parameter. Pulling beams into the BHoM, the Location of the IFramingElement is by default the centre centre location line of the element. The idea is to create a method querying an IFramingElement of potential "z Justification" parameter value and return the location line accordingly, matching the defining line in Revit.

### PR DESCRIPTION
**Pull request by annastigenberg**
### NOTE: Depends on 

https://github.com/BHoM/BHoM_Engine/pull/2118/commits
   
### Issues addressed by this PR

Closes #894

### Test files

[Issue#894.zip](https://github.com/BHoM/Revit_Toolkit/files/5494021/Issue.894.zip)

### Additional comments

Creating a method querying an IFramingElement of the "z Justification" Revit parameter and returning the in Revit geometry defining location line. Uses the methods created in PR #2118, in the Physical_Engine.

### CODE REVIEW

File: Revit_Engine/Query/RevitDefiningLine.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            {
+                definingLine = element.TopCentreline();
+                return definingLine;
+            }
+            else if (zJustification == "Bottom")
+            {
+                definingLine = element.BottomCentreline();
+                return definingLine;
+            }
+            else if (zJustification == "Origin" || zJustification =="Center")
```
Comment (FraserGreenroyd) [inline Revit_Engine/Query/RevitDefiningLine.cs:67]:
```suggestion
            else if (zJustification == "Origin" || zJustification == "Center")
```

File: Revit_Engine/Query/RevitDefiningLine.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        [Description("Returns the in Revit defining location line of an IFramingElement, based on its z Justification Revit parameter.")]
+        [Input("element", "The IFramingElement to query the in Revit defining location line of.")]
+        [Output("curve", "The in Revit geometry defining location line of the element.")]
+        public static ICurve RevitDefiningLine(this IFramingElement element)
+        {
+            string zJustification = element.GetRevitParameterValue("z Justification").ToString();
+
+            ICurve definingLine = null;
+
+            if (zJustification == "Top")
```
Comment (FraserGreenroyd) [inline Revit_Engine/Query/RevitDefiningLine.cs:57]: Can the `zJustification` ever be set to `top` rather than `Top`? Cause if so then checking only `Top` would render all instances of `top` to fall through.

It might be worth while doing `zJustification = zJustification.ToLower();` to force the justification to be lowercase, then analyse the options as `top`, `bottom`, etc. ?

File: Revit_Engine/Revit_Engine.csproj [hunk 3]
```diff
@@ -85,6 +85,9 @@
       <HintPath>C:\ProgramData\BHoM\Assemblies\Library_Engine.dll</HintPath>
       <Private>False</Private>
     </Reference>
+    <Reference Include="Physical_Engine">
+      <HintPath>..\..\..\ProgramData\BHoM\Assemblies\Physical_Engine.dll</HintPath>
```
Comment (FraserGreenroyd) [inline Revit_Engine/Revit_Engine.csproj:89]: I'll get BHoMBot to run a report for you to identify issues with the CSProject file :smile:

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
This will need to wait while [this PR](https://github.com/BHoM/BHoM_Engine/pull/2118) is closed out because the branch names aren't correct.

Also some comments for this code as well.

Happy to have a chat @annastigenberg about the linking of pull requests when doing dependant work, but hopefully [this page on branch naming conventions](https://github.com/BHoM/documentation/wiki/Resolving-an-Issue#branch-naming-conventions) might help initially :smile:

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project compliance

File: Revit_Engine/Query/RevitDefiningLine.cs (hunk 2 above)
Comment (annastigenberg) [reply Revit_Engine/Query/RevitDefiningLine.cs:57]: It must be starting with uppercase, and Revit seem to automatically correct a manual typing of lower case (at least in the structural template), but could add this anyways to be safe.

File: Revit_Engine/Query/RevitDefiningLine.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            }
+            else if (zJustification == "bottom")
+            {
+                definingLine = element.BottomCentreline();
+                return definingLine;
+            }
+            else if (zJustification == "origin" || zJustification == "center")
+            {
+                return element.Location;
+            }
```
Comment (FraserGreenroyd) [inline Revit_Engine/Query/RevitDefiningLine.cs:72]: I just realised this variable isn't really needed because we're not doing any subsequent analysis of the line, just returning it from the element. So we could remove this, and simplify the code to look like this:

```suggestion
            if (zJustification == "top")
                return element.TopCentreline();
            else if (zJustification == "bottom")
                return element.BottomCentreline();
            else if (zJustification == "origin" || zJustification == "center")
                return element.Location;
```

What do you think? This is just a suggestion so making it a PR comment rather than a requested change :smile:

Comment (annastigenberg) [reply Revit_Engine/Query/RevitDefiningLine.cs:72]: Yes agree!

**Comment by FraserGreenroyd** [comment]
@annastigenberg how are you doing with the changes needed, do you need a hand or all ok? :smile:

**Comment by annastigenberg** [comment]
@FraserGreenroyd Pushed just now!

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project compliance

**Comment by annastigenberg** [comment]
@BHoMBot fix project file ref. 1395038893

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project compliance

**Review by FraserGreenroyd — APPROVED** [review]
This all LGTM so I'm happy - great work @annastigenberg :smile:

**Review by pawelbaran — CHANGES_REQUESTED** [review]
I have read the code and tested the functionality, the concept looks good 👍 However, the code requires a few improvements to make it fully robust and clean - please see the code comments.

Regarding the changes in .csproj file - @FraserGreenroyd I understand the update was made by BHoMBot, but it is impossible to review them due to indentation changes. It will not be able to approve the change that cannot be read by a human being, especially that I noticed some changed being enforced to the file structure (which makes it simply dangerous IMO). Could we please try to speak to BHoMBot to fix the indentation, so that I (and others in other PRs) could clearly see what has been changed? Thanks!

**Comment by tiagogrossi** [comment]
Hello @annastigenberg :) poking my nose in this PR hehe.. are you able to look at the existing BH.Revit.Engine.Core.LocationCurve to compare with what you done here? There is an overload specific to `IFramingElement`, and id it's not working as it should maybe worth fixing it? Or even create a new overload? @pawelbaran  thoughts?

**Comment by pawelbaran** [comment]
This query happens after the object gets pulled and its location curve extracted @tiagogrossi. It is more of a supplement to the core functionality: the convert works fine acc. to the philosophy that we follow, this PR is meant to make it easy to switch between Revit vs BHoM approach.

File: Revit_Engine/Query/RevitDefiningLine.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        {
+            string zJustification = element.GetRevitParameterValue("z Justification").ToString();
+
+            zJustification = zJustification.ToLower();
+
+            if (zJustification == "top")
+                return element.TopCentreline();
+            else if (zJustification == "bottom")
+                return element.BottomCentreline();
+            else if (zJustification == "origin" || zJustification == "center")
```
[the remark this answers is not in the document]
Comment (annastigenberg) [reply Revit_Engine/Query/RevitDefiningLine.cs:61]: Yes that could be useful. Will create a pull request in the BHoM_Engine for that.

[the remark this answers is not in the document]
Comment (FraserGreenroyd) [reply Revit_Engine/Query/RevitDefiningLine.cs:61]: @annastigenberg can we close this PR out first unless it's gonna depend on it, just to save from having PRs hanging open for too long? 😄

[the remark this answers is not in the document]
Comment (annastigenberg) [reply Revit_Engine/Query/RevitDefiningLine.cs:61]: Yes this PR will then be dependent on that new PR. But I am creating that new PR now so hopefully it will not be too long. :)

[the remark this answers is not in the document]
Comment (annastigenberg) [reply Revit_Engine/Query/RevitDefiningLine.cs]: A new PR has been created with the same branch name as this one, and this PR has a new commit adding the new method. @pawelbaran @FraserGreenroyd

File: Revit_Engine/Query/ZJustifiedCurve.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+namespace BH.Engine.Adapters.Revit
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("Returns the driving curve of the given IFramingElement adjusted based on the 'z Justification' parameter value pulled from Revit.")]
```
Comment (pawelbaran) [inline Revit_Engine/Query/ZJustifiedCurve.cs:37]: I would extend the description a bit to make it super clear:
```suggestion
        [Description("Returns the driving curve of the given IFramingElement adjusted based on the 'z Justification' parameter value pulled from Revit. Important: neither z offset nor any y justification is taken into account.")]
```

File: Revit_Engine/Query/ZJustifiedCurve.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        [Input("element", "The IFramingElement to query the in Revit defining location line of.")]
+        [Output("curve", "The in Revit geometry defining location line of the element.")]
+        public static ICurve ZJustifiedCurve(this IFramingElement element)
+        {
+            string zJustification = element.GetRevitParameterValue("z Justification").ToString();
+
+            zJustification = zJustification.ToLower();
```
Comment (pawelbaran) [inline Revit_Engine/Query/ZJustifiedCurve.cs:44]: You need extra null checks here to make sure the code will not crash:
```suggestion
            if (element == null)
                return null;
                
            string zJustification = element.GetRevitParameterValue("z Justification")?.ToString();
            zJustification = zJustification?.ToLower();
```

File: Revit_Core_Adapter/Revit_Core_Adapter.csproj [hunk 8]
```diff
@@ -12,6 +12,7 @@
     <AssemblyName>Revit_Core_Adapter_2018</AssemblyName>
     <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
     <FileAlignment>512</FileAlignment>
+    <TargetFrameworkProfile />
```
Comment (pawelbaran) [inline Revit_Core_Adapter/Revit_Core_Adapter.csproj:15]: Please revert the changes in this file, it is irrelevant to this PR (you can e.g. switch back to `master` for a second, copy the file somewhere, then switch back to this branch and paste the file).

File: Revit_Core_Engine/Revit_Core_Engine.csproj [hunk 9]
```diff
@@ -12,6 +12,7 @@
     <AssemblyName>Revit_Core_Engine_2018</AssemblyName>
     <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
     <FileAlignment>512</FileAlignment>
+    <TargetFrameworkProfile />
```
Comment (pawelbaran) [inline Revit_Core_Engine/Revit_Core_Engine.csproj:15]: Same as above: please revert the changes in this file, it is irrelevant to this PR

**Review by pawelbaran — CHANGES_REQUESTED** [review]
I took another round of review: the code looks much better now, I would only add some null checks and more descriptions - please see the code comments.

Besides that, the .csproj files got broken, which I believe was not your intention, but rather BHoMBot's frenzy, am I correct?

**Comment by annastigenberg** [comment]
Recreating PR due to error with unwanted change of files.
